File: samples/dotnet/museum-exhibit-studio/MuseumExhibitService.cs
using GitHub.Copilot;

namespace MuseumExhibitStudio;

public sealed record GeneratedExhibit(string Content, ExhibitValidation Validation);

public sealed class MuseumExhibitService(ICuratorClient client)
{
    public static readonly TimeSpan GenerationTimeout = TimeSpan.FromMinutes(2);

    public async Task<GeneratedExhibit> GenerateAsync(
        IEnumerable<string> approvedFacts,
        string? model = null,
        CancellationToken cancellationToken = default)
    {
        var prompt = CuratorPrompts.BuildExhibitPrompt(approvedFacts);

        await client.StartAsync(cancellationToken);
        try
        {
            await using var session = await client.CreateSessionAsync(
                CreateSessionConfiguration(model),
                cancellationToken);
            var content = await session.SendAndWaitAsync(
                prompt,
                GenerationTimeout,
                cancellationToken);

            if (string.IsNullOrWhiteSpace(content))
            {
                throw new InvalidOperationException("The curator returned no exhibit content.");
            }

            return new GeneratedExhibit(content, ExhibitValidator.Validate(content));
        }
        finally
        {
            await client.StopAsync();
        }
    }

    public static SessionConfig CreateSessionConfiguration(string? model = null) => new()
    {
        ClientName = "museum-exhibit-studio",
        Model = string.IsNullOrWhiteSpace(model) ? null : model,
        AvailableTools = [],
        Streaming = false,
        SystemMessage = new SystemMessageConfig
        {
            Mode = SystemMessageMode.Replace,
            Content = CuratorPrompts.SystemMessage
        }
    };
}

File: samples/python/museum-exhibit-studio/museum_exhibit_service.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from curator_prompts import SYSTEM_MESSAGE, build_exhibit_prompt
from exhibit_validator import ExhibitValidation, validate_exhibit

GENERATION_TIMEOUT_SECONDS = 120.0


@dataclass(frozen=True)
class GeneratedExhibit:
    content: str
    validation: ExhibitValidation


def create_session_configuration(model: str | None = None) -> dict[str, Any]:
    return {
        "client_name": "museum-exhibit-studio",
        "model": model.strip() if model and model.strip() else None,
        "available_tools": [],
        "streaming": False,
        "system_message": {"mode": "replace", "content": SYSTEM_MESSAGE},
    }


class MuseumExhibitService:
    def __init__(self, client: Any) -> None:
        self._client = client

    async def generate(
        self, approved_facts: list[str] | tuple[str, ...], model: str | None = None
    ) -> GeneratedExhibit:
        prompt = build_exhibit_prompt(approved_facts)
        session = None
        try:
            await self._client.start()
            session = await self._client.create_session(
                **create_session_configuration(model)
            )
            response = await session.send_and_wait(
                prompt, timeout=GENERATION_TIMEOUT_SECONDS
            )
            content = getattr(getattr(response, "data", None), "content", None)
            if not content or not content.strip():
                raise RuntimeError("The curator returned no exhibit content.")
            return GeneratedExhibit(content, validate_exhibit(content))
        finally:
            try:
                if session is not None:
                    await session.disconnect()
            finally:
                await self._client.stop()
